thetis/shallowwater_eq.py

```diff
@@ -998,3 +1001,455 @@ def residual(self, label, solution, solution_old, fields, fields_old, bnd_condit
         eta = fields['eta']
         eta_old = fields_old['eta']
         return self.residual_uv_eta(label, uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions)
+
+
+class ShallowWaterMomentumResidualTerm(ShallowWaterTerm):
+    """
+    Generic term in the strong form shallow water momentum equation.
+    """
+
+    def __init__(self, u_space, eta_space,
+                 bathymetry=None,
+                 options=None):
+        super(ShallowWaterMomentumResidualTerm, self).__init__(u_space, bathymetry, options)
+
+        self.options = options
+
+        self.u_space = u_space
+        self.eta_space = eta_space
+
+        self.u_continuity = element_continuity(self.u_space.ufl_element()).horizontal
+        self.eta_is_dg = element_continuity(self.eta_space.ufl_element()).horizontal == 'dg'
+
+
+class ShallowWaterContinuityResidualTerm(ShallowWaterTerm):
+    """
+    Generic term in the strong form depth-integrated continuity equation.
+    """
+    def __init__(self, eta_space, u_space,
+                 bathymetry=None,
+                 options=None):
+        super(ShallowWaterContinuityResidualTerm, self).__init__(eta_space, bathymetry, options)
+
+        self.eta_space = eta_space
+        self.u_space = u_space
+
+        self.u_continuity = element_continuity(self.u_space.ufl_element()).horizontal
+        self.eta_is_dg = element_continuity(self.eta_space.ufl_element()).horizontal == 'dg'
+
+
+class ExternalPressureGradientResidual(ShallowWaterMomentumResidualTerm):
+    r"""
+    External pressure gradient term, :math:`g \nabla \eta`
+    """
+    name = 'ExternalPressureGradient'
+
+    def residual_cell(self, uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions):
+
+        f = g_grav * grad(eta)
+
+        return -f
+
+    def residual_edge(self, uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions):
+        raise NotImplementedError  # FIXME
+
+
+class HUDivResidual(ShallowWaterContinuityResidualTerm):
```

Comment:
Writing a separate residual form for each term looks like a lot of duplicated code. I wonder if there's a better way to do this (maybe not?).